models/classes/runner/RunnerToolStates.php

```diff
@@ -0,0 +1,105 @@
+<?php
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2017 (original work) Open Assessment Technologies SA ;
+ */
+
+namespace oat\taoQtiTest\models\runner;
+
+use oat\oatbox\service\ServiceManager;
+
+trait RunnerToolStates
+{
+    /**
+     * @param string $name
+     * @return string
+     */
+    abstract public function hasRequestParameter($name);
+
+    /**
+     * @param string $name
+     * @return string mixed
+     */
+    abstract public function getRequestParameter($name);
+
+    /**
+     * @return string
+     */
+    abstract public function getRawRequestParameter();
+
+    /**
+     * @return ServiceManager
+     */
+    abstract public function getServiceLocator();
+
+    /**
+     * @return RunnerService
+     */
+    abstract public function getRunnerService();
+
+    /**
+     * @return RunnerServiceContext
+     */
+    abstract public function getServiceContext();
+
+    /**
+     * Save the tool state if some are found in the current request
+     *
+     * @throws \oat\oatbox\service\exception\InvalidServiceManagerException
+     * @throws \common_Exception
+     *
+     * @return boolean true if some state was found and saved
+     */
+    protected function saveToolStates()
+    {
+        // since the parameter content is a JSON string
+        // we need to load it using the raw mode
+        $param = $this->getRawRequestParameter('toolStates');
```

Comment:
If the parameter doesn't exist an exception + 500 is thrown.
The `toolStates` parameter is optional, if no tool has a state to post we don't post anything